update: Added basic methods for data augmentation

vidriduch opened the pull request:
- realtime_augmentation.py 
  - added method for bluring entire batch of entire batch of images. Robot is constantly moving and many images are going to be blured
  - added fancy PCA color augmentation based on http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf and discussion here: https://groups.google.com/forum/#!topic/lasagne-users/meCDNeA9Ud4
  - added method which flips entire batch of images and their masks horizontaly
- utils.py - added debug windows for methods in realtime_augmentation.py
- obstacle_adder.py - toolkit program for artificially adding obstacles
                    to our image dataset
- obstacles_cropped - added 25 cropped obstacles extracted from
                    http://dai.fmph.uniba.sk/projects/smelyzajko/robotour/data/sad_janka_krala.zip
- with_obs - folder with some example of added obstacles to some images

vidriduch committed:
update: Added basic methods for data augmentation

* realtime_augmentation.py - added method for bluring entire batch of
                             entire batch of images. Robot is constatly
                             moving and many images are going to be
                             blured

                           - added fancy PCA color augmentation based on
                             http://www.cs.toronto.edu/~fritz/absps/imagenet.pdf
                             and discussion https://groups.google.com/forum/#!topic/lasagne-users/meCDNeA9Ud4

                           - added method whitch flips entire batch of
                             images and their masks horizontaly

* utils.py - added debug windows for methods in realtime_augmentation.py

* obstacle_adder.py - toolkit program for artificially adding obstacles
                      to our image dataset

* obstacles_cropped - added 25 cropped obstacles extracted from
                      http://dai.fmph.uniba.sk/projects/smelyzajko/robotour/data/sad_janka_krala.zip

* with_obs - folder with some example of added obstacles to some images

vidriduch committed:
update: added new cropped obstacles and images

* Added new obstacles from plzen dataset

* Added new pictures with artificially added obstacles

* obstacle_adder - change filename generation from id to random 32bit
int

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,125 @@
+                    obs = cv.flip(obs, 1)
+                    obs_mask = cv.flip(obs_mask, 1)
+                if k == ord('s'):
+                    filename = './with_obs/{}_obs.png'.format(getrandbits(32))
```
@Foxiy Just wanted to point this out in your previous commit =)

mrshu commented on realtime_augmentation.py:
```diff
@@ -0,0 +1,30 @@
+import cv2 as cv
```
@Foxiy should this be in utils?

vidriduch replied:
@mrshu I just didn't want to clutter up the utils file with these augmentation functions. But you might be right and probably should be merged with utils

mrshu commented on utils.py:
```diff
@@ -76,13 +109,40 @@ def dataset_from_folder(folder, img_ext='.png', mask_ext='.trn'):
+    jitter_data = ra.add_color_noise(data, eigenvalues, pc)
+    pca_data = np.clip(jitter_data * 255, 0, 255).astype('uint8')
+
+    flipped_data, flipped_mask = ra.flip_batch(data, data_mask)
```
@Foxiy all of these things could probably happen as part of `dataset_from_folder`. The only problem I see there is that `flip_batch` works on batches (obviously) while `dataset_from_folder` reads (and yields out) the data in a sequential manner.

What do you think?

vidriduch replied:
@mrshu I think this is better because its flips entire batch at the same time and numpy takes care of all the 'loop' stuff which is faster then doing this in python sequentially. Also I think we should stick to batches because they going to be loaded to memory and if we would do it on entire dataset we might run into memory problems latter on.

vidriduch committed:
update: augmentation fncts moved to utils

* realtime_augmaentation.py - functions moved to utils.py and file removed

* obstacle_adder.py - function numpy_to_trn moved also moved to utils.py

vidriduch commented:
@mrshu Thank you for review I moved those function to utils.py file. If you could give it one more look I would greatly appreciate it

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,118 @@
+    return np.asarray(obstacles), np.asarray(obstacles_mask)
+
+
+def __insert_image(image, mask, x, y, h, w):
```
@Foxiy is there really a need for `__` in the definition of this function?

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,118 @@
+
+
+def __insert_image(image, mask, x, y, h, w):
+    global obs_mask, obs
```
@Foxiy Rather than doing this globally, do you think it would make sense to just wrap this in an object? If that would lead to too much trouble do not even think about it.

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,118 @@
+    return image[h:-h, w:-w], mask[h:-h, w:-w]
+
+
+def visualize_labels(img, mask):
```
@Foxiy This looks like something that could be part of utils.py.

I am also not a huge fan of calling `cv.imshow` from functions that might be imported from random places. If that call is really unavoidable, we should at least let the caller change the title of the frame.

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,118 @@
+
+# mouse callback function
+def place(event, x, y, flags, param):
+    global img, mask, obs, obs_mask, new_img, new_mask
```
@Foxiy would love to get rid of these by making the whole thing an object.

mrshu commented on obstacle_adder.py:
```diff
@@ -0,0 +1,118 @@
+        for i, obstacle in enumerate(obstacles):
+            obs = obstacle
+            obs_mask = obstacles_mask[i]
+            while(1):
```
@Foxiy this is a very small nitpick but I would personally prefer 

``` python
while True:
```

mrshu commented on utils.py:
```diff
@@ -16,6 +16,14 @@ def trn_to_numpy(filename):
         return numpy_array.reshape(img.shape[0], img.shape[1])
 
 
+def numpy_to_trn(filename, mask):
```
@Foxiy I believe this should just return the contents of the `.trn` file (i.e. the line 23). Saving the file should probably happen somewhere else.

mrshu commented on utils.py:
```diff
@@ -16,6 +16,14 @@ def trn_to_numpy(filename):
+def numpy_to_trn(filename, mask):
+    filename_trn = "{}.trn".format(filename)
+    _, mask = cv.threshold(mask, 1, 1, cv.THRESH_BINARY)
+    f = open(filename_trn, 'w+')
```
@Foxiy It would be great if you could use the following:

``` python
with open(filename_trn, 'w+') as f:
    f.write(' '.join(str(e) for e in mask.flatten().tolist()))
```

mrshu commented on utils.py:
```diff
@@ -48,6 +56,10 @@ def visualize_proba(proba, shape, color=(0, 0, 255)):
     return img
 
 
+def visualize_mask(image, mask):
```
@Foxiy I believe the name of this function is off as it does not visualize anything as much as I can see.

mrshu commented on utils.py:
```diff
@@ -62,6 +74,56 @@ def load_image_for_dataset(filename):
+    blured_batch = []
+    for image in batch:
+        gray_im = cv.cvtColor(image, cv.COLOR_BGR2GRAY)
+        blur_amount = cv.Laplacian(gray_im, cv.CV_64F).var()/1000.
```
@Foxiy I nearly missed the `1000.` on this line, I believe that for the sake of clarity it should really say `1000.0`.

mrshu commented on utils.py:
```diff
@@ -76,13 +138,40 @@ def dataset_from_folder(folder, img_ext='.png', mask_ext='.trn'):
 
 if __name__ == '__main__':
-    for datum in dataset_from_folder('./plzen'):
+    data, data_mask = load_dataset('./plzen')
```
@Foxiy I really do not like the fact that in the actual usecase we would read the whole dataset twice, just to run augmentation in batches. 

The big win of `dataset_from_folder` function's use of `yield` is in that if things get to be processed sequentially, this is extra fast and effective.

I agree though that for our usecase (actually trying to learn some models from the augmented data) it would be best to have the whole thing in memory and just doing batch operations on it.

Do you think we could have a function called `load_datase_from_folder` that would do just that (I mean load the data from disk to memory and do augmentation)? If someone did not want augmentation, they could just pass `augmentation=False` as a keyword parameter.

Ditching the `dataset_from_folder` function is totally on the table too. I do not want to invest time into something that will not be used after all, but I really hate when things are being done twice for no apparent reason.

vidriduch committed:
update: rewriten obstacle_adder to object

* obstacle_adder.py - obstacle_adder is now ObstacleAdder object
                    - moved function visualize label to utils as
                      convenience function
                    - s-button now ask for content of .trn file and then
                      saves it rather then before calling a function to save it

* utils.py - numpy_to_trn now returns content of .trn file rather then
             saving it
           - visualize labels now returns only content of the
             window(image)
           - renamed function load_dataset to load_dataset_as_numpy

vidriduch commented:
@mrshu Thanks for the review. I tried to fixed most of the stuff. I believe this is ready for another pass. I still think that those augmentation function should work on batches since this way it can be latter use as callback to Keras train function to realtime augment batches of images rather then loading it all at once to memory.

mrshu commented:
@Foxiy roger that, will leave this to your judgment =)
